src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -988,14 +985,47 @@ where
     }
 }
 
+enum PgStreamInner<S> {
+    /// Used for the intermediate state when converting from unencrypted to ssl stream.
+    Placeholder,
+    /// An unencrypted stream.
+    Unencrypted(S),
+    /// An ssl stream.
+    Ssl(SslStream<S>),
+}
+
 /// Wraps a byte stream and read/write pg messages.
+///
+/// Cloning a `PgStream` will share the same stream but a fresh & independent write buffer,
+/// so that it can be used to write messages concurrently without interference.
 pub struct PgStream<S> {
     /// The underlying stream.
-    stream: Option<S>,
+    stream: Arc<Mutex<PgStreamInner<S>>>,
```

Comment:
The overhead for this `Mutex` will be minimal since there won't be any contention during actual use.